File: Python2/merge_two_binary_trees.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def mergeTrees(self, t1, t2):
        """
        :type t1: TreeNode
        :type t2: TreeNode
        :rtype: TreeNode
        """
        
        if t1 is None and t2 is not None:
            return t2
        if t1 is not None and t2 is None:
            return t1

        # use level-order traversal
        import Queue
        q = Queue.Queue()
        q.put([t1, t2])
        while q.qsize() != 0:
            this_nodes = q.get()
            if this_nodes[0] is None and this_nodes[1] is None:
                continue
            if this_nodes[0] is not None:
                this_nodes[0].val += (this_nodes[1].val if this_nodes[1] is not None else 0)

            # left child
            next_left_child_2 = this_nodes[1].left if this_nodes[1] is not None else None
            if next_left_child_2 is not None and this_nodes[0].left is None:
                next_left_child_1 = TreeNode(0)
                this_nodes[0].left = next_left_child_1
            else:
                next_left_child_1 = this_nodes[0].left if this_nodes[0] is not None else None

            # right child
            next_right_child_2 = this_nodes[1].right if this_nodes[1] is not None else None
            if next_right_child_2 is not None and this_nodes[0].right is None:
                next_right_child_1 = TreeNode(0)
                this_nodes[0].right = next_right_child_1
            else:
                next_right_child_1 = this_nodes[0].right if this_nodes[0] is not None else None

            q.put([next_left_child_1, next_left_child_2])
            q.put([next_right_child_1, next_right_child_2])

        return t1


if __name__ == '__main__':
    t1 = TreeNode(1)
    t1.left = TreeNode(3)
    t1.right = TreeNode(2)
    t1.left.left = TreeNode(5)
    t2 = TreeNode(2)
    t2.left = TreeNode(1)
    t2.right = TreeNode(3)
    t2.left.right = TreeNode(4)
    t2.right.right = TreeNode(7)

    sol = Solution()
    print sol.mergeTrees(t1, t2)

File: C++/merge_two_binary_trees.cpp
#include "./merge_two_binary_trees.h"
#include <queue>
#include <iostream>

using namespace std;

TreeNode* Solution::mergeTrees(TreeNode* t1, TreeNode* t2) {
    if (t1 == NULL && t2 != NULL) return t2;
    if (t1 != NULL && t2 == NULL) return t1;
    // use level-order traversal
    queue<TreeNode*> q1;
    q1.push(t1);
    q1.push(t2);
    TreeNode* pt1 = NULL;
    TreeNode* pt2 = NULL;
    while (q1.size() > 0) {
        pt1 = q1.front();
        q1.pop();
        pt2 = q1.front();
        q1.pop();
        if (pt1 != NULL && pt2 != NULL) {
            pt1->val += pt2->val;
        }
        if (pt2 != NULL) {
            if (pt1->left == NULL && pt2->left != NULL) {
                pt1->left = new TreeNode(0);
            }
            if (pt1->right == NULL && pt2->right != NULL) {
                pt1->right = new TreeNode(0);
            }
        }
        
        if (pt1 != NULL || pt2 != NULL) {
            q1.push(pt1 != NULL ? pt1->left : NULL);
            q1.push(pt2 != NULL ? pt2->left : NULL);
            q1.push(pt1 != NULL ? pt1->right : NULL);
            q1.push(pt2 != NULL ? pt2->right : NULL);
        }
    }
    return t1;
}


int main() {
    TreeNode t1(1);
    t1.left = new TreeNode(3);
    t1.right = new TreeNode(2);
    t1.left->left = new TreeNode(5);
    Utility::traverseTreeByLevelOrder(&t1);

    TreeNode t2(2);
    t2.left = new TreeNode(1);
    t2.right = new TreeNode(3);
    t2.left->right = new TreeNode(4);
    t2.right->right = new TreeNode(7);
    Utility::traverseTreeByLevelOrder(&t2);

    Solution* sol = new Solution();
    TreeNode* output = sol->mergeTrees(&t1, &t2);
    Utility::traverseTreeByLevelOrder(output);
    return 0;
}